src/coprocessor/codec/convert.rs

```diff
@@ -43,53 +101,163 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
     if val > upper_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
-/// `convert_float_to_uint` converts a f64 value to a u64 value.
-/// Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_uint(fval: f64, upper_bound: u64, tp: FieldTypeTp) -> Result<u64> {
-    // TODO any performance problem to use round directly?
+/// Converts a byte arrays to an i64 in best effort.
+pub fn convert_bytes_to_int(ctx: &mut EvalContext, bytes: &[u8], tp: FieldTypeTp) -> Result<i64> {
+    let s = match str::from_utf8(bytes) {
+        Ok(s) => s,
+        Err(err) => {
+            ctx.handle_truncate(true)?;
+            let (valid, _) = bytes.split_at(err.valid_up_to());
+            unsafe { str::from_utf8_unchecked(valid) }
+        }
+    };
+    let s = s.trim();
+    let vs = get_valid_int_prefix(ctx, s)?;
+    let val = vs.parse::<i64>();
+    match val {
+        Ok(val) => convert_int_to_int(ctx, val, tp),
+        Err(_) => {
+            ctx.handle_overflow(Error::overflow("BIGINT", &vs))?;
+            let val = if vs.starts_with('-') {
+                integer_signed_lower_bound(tp)
+            } else {
+                integer_signed_upper_bound(tp)
+            };
+            Ok(val)
+        }
+    }
+}
+
+/// Converts a `DateTime` to an i64 value
+#[inline]
+pub fn convert_datetime_to_int(
+    ctx: &mut EvalContext,
+    dt: &DateTime,
+    tp: FieldTypeTp,
+) -> Result<i64> {
+    // TODO: avoid this clone after refactor the `Time`
+    let mut t = dt.clone();
+    t.round_frac(DEFAULT_FSP)?;
+    let val = t.to_decimal()?.as_i64_with_ctx(ctx)?;
+    convert_int_to_int(ctx, val, tp)
+}
+
+/// Converts a `Duration` to an i64 value
+#[inline]
+pub fn convert_duration_to_int(
+    ctx: &mut EvalContext,
+    dur: Duration,
+    tp: FieldTypeTp,
+) -> Result<i64> {
+    // It's OK to clone because duration only occupies 8 bytes after #4858 merged
+    let dur = dur.round_frac(DEFAULT_FSP)?;
+    let val = Decimal::try_from(dur)?.as_i64_with_ctx(ctx)?;
+    convert_int_to_int(ctx, val, tp)
+}
+
+/// Converts a `Decimal` to an i64 value
+#[inline]
+pub fn convert_decimal_to_int(
+    ctx: &mut EvalContext,
+    dec: &Decimal,
+    tp: FieldTypeTp,
+) -> Result<i64> {
+    // TODO: avoid this clone
+    let dec = match dec.clone().round(0, RoundMode::HalfEven) {
+        Res::Ok(d) => d,
+        Res::Overflow(d) => {
+            ctx.handle_overflow(Error::overflow("DECIMAL", ""))?;
+            d
+        }
+        Res::Truncated(d) => {
+            ctx.handle_truncate(true)?;
+            d
+        }
+    };
+    let val = dec.as_i64_with_ctx(ctx)?;
+    convert_int_to_int(ctx, val, tp)
+}
+
+/// Converts a `Json` to an i64 value
+#[inline]
+pub fn convert_json_to_int(ctx: &mut EvalContext, json: &Json, tp: FieldTypeTp) -> Result<i64> {
+    let val = json.cast_to_int();
```

Comment:
`Json::cast_to_int` implementation looks to be different as TiDB's `ConvertJSONToInt`.. More specifically, when JSON type is String, in TiDB `StrToInt` is used while in TiKV there is only simply `s.parse::<i64>().unwrap_or(0)`.

Reply:
The all `Json` cast functions have been fixed in #4987, See: https://github.com/tikv/tikv/pull/4987/files#diff-35e557441edd7ec066b70047ba17f8b2R25